File: fangzhen/matlab/draw_pid_diagram.m
%% PID控制原理图绘制
%  生成 PID 控制原理框图, 替换原 pid.png

clear; close all;

fig = figure('Color','w','Position',[100,100,1000,420], 'Visible','off');
hold on; axis equal; axis off;

% 设置中文字体以避免渲染警告
set(fig, 'DefaultTextFontName', 'SimHei');
set(fig, 'DefaultAxesFontName', 'SimHei');

% ---- 颜色与样式 ----
box_color   = [0.90, 0.95, 1.00];   % 方框填充色 (浅蓝)
box_edge    = [0.00, 0.30, 0.60];   % 方框边框色
sum_color   = [0.95, 0.95, 0.95];   % 求和点填充色
arrow_color = [0.15, 0.15, 0.15];   % 箭头/连线颜色
text_color  = [0.00, 0.00, 0.00];   % 文本颜色

fs_label = 10;    % 标签字号
fs_title = 11;    % 标题字号
fs_math  = 11;    % 公式字号

% ---- 坐标布局 ----
% 节点 x 坐标
x_ref   = 1.0;    % 参考输入
x_sum1  = 3.0;    % 比较点
x_pid   = 5.5;    % PID 控制器
x_sum2  = 8.0;    % PID 内部求和
x_plant = 10.5;   % 被控对象
x_out   = 13.0;   % 输出

% 节点 y 坐标
y_main  = 5.0;    % 主通路
y_p     = 7.2;    % P 分支
y_i     = 5.0;    % I 分支
y_d     = 2.8;    % D 分支
y_fb_up = 7.5;    % 反馈线上端
y_fb_lo = 7.5;    % 反馈线下端 (同)

bx_w = 1.2;       % 方框半宽
bx_h = 0.55;      % 方框半高

%% ---- 1. 方框 ----
% 比较点圆圈
draw_circle(x_sum1, y_main, 0.22, sum_color, arrow_color);

% PID 三个分支框
draw_box(x_pid, y_p, bx_w, bx_h, box_color, box_edge);  % P
draw_box(x_pid, y_i, bx_w, bx_h, box_color, box_edge);  % I
draw_box(x_pid, y_d, bx_w, bx_h, box_color, box_edge);  % D

% PID 内部求和点
draw_circle(x_sum2, y_main, 0.22, sum_color, arrow_color);

% 被控对象方框
draw_box(x_plant, y_main, bx_w, bx_h, [0.95, 0.90, 0.85], [0.60, 0.40, 0.10]);

% 传感器方框
draw_box(x_plant, y_fb_lo, 0.55, 0.40, sum_color, arrow_color);

%% ---- 2. 框内文字 ----
% PID 分支
text(x_pid, y_p,   ' $K_p$',       'HorizontalAlignment','center','FontSize',fs_math+2,'Interpreter','latex');
text(x_pid, y_i,   ' $K_i\!\int$', 'HorizontalAlignment','center','FontSize',fs_math+2,'Interpreter','latex');
text(x_pid, y_d,   ' $K_d\frac{d}{dt}$','HorizontalAlignment','center','FontSize',fs_math+2,'Interpreter','latex');

% 被控对象
text(x_plant, y_main, '被控对象','HorizontalAlignment','center','FontSize',fs_title,'Interpreter','latex');

% 传感器
text(x_plant, y_fb_lo, '传感器','HorizontalAlignment','center','FontSize',fs_label,'Interpreter','latex');

%% ---- 3. 主通路连线 ----
% 参考输入 r(t) → 比较点
draw_arrow(x_ref, y_main, x_sum1-0.22, y_main, arrow_color);
text(1.7, y_main+0.35, '$r(t)$', 'FontSize',fs_math,'Interpreter','latex','Color',text_color);
text(0.85, y_main-0.80, '参考输入','FontSize',fs_label,'Interpreter','latex','HorizontalAlignment','center');

% 比较点 → PID 分叉点 (x=4.0)
x_fork = 4.0;
draw_arrow(x_sum1+0.22, y_main, x_fork, y_main, arrow_color);
text(3.3, y_main+0.35, '$e(t)$', 'FontSize',fs_math,'Interpreter','latex','Color',text_color);

% 分叉 → P
draw_line(x_fork, y_main, x_fork, y_p, arrow_color);
draw_arrow(x_fork, y_p, x_pid-bx_w, y_p, arrow_color);

% 分叉 → I
draw_line(x_fork, y_main, x_pid-bx_w, y_main, arrow_color);
% 在分叉处做标记
plot(x_fork, y_main, 'ko','MarkerSize',4,'MarkerFaceColor',arrow_color);

% 分叉 → D
draw_line(x_fork, y_main, x_fork, y_d, arrow_color);
draw_arrow(x_fork, y_d, x_pid-bx_w, y_d, arrow_color);

% 分叉标记
text(x_fork-0.15, y_main-0.30, '$+$','FontSize',fs_math-1,'Interpreter','latex');

% P → 求和点
draw_arrow(x_pid+bx_w, y_p, x_sum2-0.22, y_p, arrow_color);
draw_line(x_sum2-0.22, y_p, x_sum2-0.22, y_main+0.22, arrow_color);
plot(x_sum2-0.22, y_p, 'ko','MarkerSize',3,'MarkerFaceColor',arrow_color);

% I → 求和点
draw_arrow(x_pid+bx_w, y_i, x_sum2-0.22, y_i, arrow_color);

% D → 求和点
draw_arrow(x_pid+bx_w, y_d, x_sum2-0.22, y_d, arrow_color);
draw_line(x_sum2-0.22, y_d, x_sum2-0.22, y_main-0.22, arrow_color);
plot(x_sum2-0.22, y_d, 'ko','MarkerSize',3,'MarkerFaceColor',arrow_color);

% 求和点 → 被控对象
draw_arrow(x_sum2+0.22, y_main, x_plant-bx_w, y_main, arrow_color);
text(8.8, y_main+0.35, '$u(t)$', 'FontSize',fs_math,'Interpreter','latex','Color',text_color);

% 被控对象 → 输出
draw_arrow(x_plant+bx_w, y_main, x_out, y_main, arrow_color);
text(12.2, y_main+0.35, '$y(t)$', 'FontSize',fs_math,'Interpreter','latex','Color',text_color);
text(x_out+0.20, y_main-0.80, '输出','FontSize',fs_label,'Interpreter','latex','HorizontalAlignment','center');

%% ---- 4. 反馈通路 ----
% 输出分叉点 (x=12.0)
x_fb = 12.0;
plot(x_fb, y_main, 'ko','MarkerSize',4,'MarkerFaceColor',arrow_color);

% 向下 → 传感器
draw_line(x_fb, y_main, x_fb, y_fb_lo, arrow_color);
draw_arrow(x_fb, y_fb_lo, x_plant+bx_w, y_fb_lo, arrow_color);

% 传感器 → 向左 → 向上 → 比较点
x_fb_left = x_plant - bx_w;
draw_arrow(x_plant-bx_w, y_fb_lo, 1.5, y_fb_lo, arrow_color);
draw_line(1.5, y_fb_lo, 1.5, y_main, arrow_color);
draw_arrow(1.5, y_main, x_sum1-0.22, y_main, arrow_color);

% 反馈标记
text(0.85, y_fb_lo-0.60, '反馈','FontSize',fs_label,'Interpreter','latex','HorizontalAlignment','center');

%% ---- 5. 比较点标注 ----
text(x_sum1-0.60, y_main-0.45, '$+$','FontSize',fs_math,'Interpreter','latex');
text(x_sum1-0.45, y_main+0.48, '$-$','FontSize',fs_math,'Interpreter','latex');

%% ---- 6. 求和点标注 ----
text(x_sum2-0.20, y_main+1.30, '$+$','FontSize',fs_math-1,'Interpreter','latex');

%% ---- 7. 虚线框: PID控制器 ----
x_pid_left  = x_fork - 0.25;
x_pid_right = x_sum2 + 0.35;
y_pid_top   = y_p + bx_h + 0.35;
y_pid_bot   = y_d - bx_h - 0.35;

plot([x_pid_left, x_pid_right, x_pid_right, x_pid_left, x_pid_left], ...
     [y_pid_top, y_pid_top, y_pid_bot, y_pid_bot, y_pid_top], ...
     'k--','LineWidth',1.0);
text(x_pid, y_pid_top+0.30, 'PID 控制器','HorizontalAlignment','center',...
     'FontSize',fs_title,'Interpreter','latex','Color',text_color);

%% ---- 8. 全局比例与保存 ----
xlim([0, 14.5]);
ylim([0, 9.5]);

out_dir = fullfile(fileparts(mfilename('fullpath')), '..', '..', ...
    'lunwen','QLULatex','QLUThesisLatexTemplate-master','Thesis','static','figures');
if ~exist(out_dir, 'dir'), mkdir(out_dir); end
exportgraphics(fig, fullfile(out_dir, 'pid.png'), 'Resolution', 200);
fprintf('PID 原理图已保存: %s\n', fullfile(out_dir, 'pid.png'));

%% ======================== 绘图辅助函数 ========================
function draw_box(cx, cy, hw, hh, fc, ec)
    % 绘制填充矩形方框
    xv = [cx-hw, cx+hw, cx+hw, cx-hw];
    yv = [cy-hh, cy-hh, cy+hh, cy+hh];
    patch(xv, yv, fc, 'EdgeColor', ec, 'LineWidth', 1.5);
end

function draw_circle(cx, cy, r, fc, ec)
    % 绘制填充圆 (求和点)
    th = linspace(0, 2*pi, 60);
    patch(cx + r*cos(th), cy + r*sin(th), fc, 'EdgeColor', ec, 'LineWidth', 1.2);
end

function draw_arrow(x1, y1, x2, y2, color)
    % 绘制带箭头的线段
    plot([x1, x2], [y1, y2], '-', 'Color', color, 'LineWidth', 1.5);
    % 箭头三角形
    len = 0.18;  ang = 0.45;
    dx = x2 - x1;  dy = y2 - y1;
    L = hypot(dx, dy);
    if L < 1e-8, return; end
    ux = dx/L;  uy = dy/L;
    ax1 = x2 - len*cos(atan2(uy,ux)-ang);
    ay1 = y2 - len*sin(atan2(uy,ux)-ang);
    ax2 = x2 - len*cos(atan2(uy,ux)+ang);
    ay2 = y2 - len*sin(atan2(uy,ux)+ang);
    patch([x2, ax1, ax2], [y2, ay1, ay2], color, 'EdgeColor', 'none');
end

function draw_line(x1, y1, x2, y2, color)
    % 绘制无箭头线段
    plot([x1, x2], [y1, y2], '-', 'Color', color, 'LineWidth', 1.5);
end

File: fangzhen/matlab/draw_pid_diagram.py
"""PID控制原理图绘制 - 使用 matplotlib"""
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import numpy as np

# ---------- 中文字体 ----------
plt.rcParams['font.sans-serif'] = ['SimHei', 'Microsoft YaHei', 'DejaVu Sans']
plt.rcParams['axes.unicode_minus'] = False

# ---------- 图形尺寸 ----------
fig, ax = plt.subplots(figsize=(12, 4.8))
ax.set_xlim(0, 14.5)
ax.set_ylim(0, 9.5)
ax.set_aspect('equal')
ax.axis('off')

# ---------- 颜色 ----------
box_color   = '#E6F0FF'
box_edge    = '#1A4D80'
sum_color   = '#F0F0F0'
arr_color   = '#2A2A2A'
dash_color   = '#555555'

# ---------- 辅助函数 ----------
def draw_box(cx, cy, w, h, fc=box_color, ec=box_edge):
    rect = mpatches.FancyBboxPatch((cx-w/2, cy-h/2), w, h,
            boxstyle='round,pad=0.06', fc=fc, ec=ec, lw=1.5)
    ax.add_patch(rect)

def draw_circle(cx, cy, r):
    circ = plt.Circle((cx, cy), r, fc=sum_color, ec=arr_color, lw=1.2, zorder=4)
    ax.add_patch(circ)

def draw_arrow(x1, y1, x2, y2, lw=1.5):
    ax.annotate('', xy=(x2, y2), xytext=(x1, y1),
        arrowprops=dict(arrowstyle='->', color=arr_color, lw=lw))

def draw_line(x1, y1, x2, y2, lw=1.5):
    ax.plot([x1, x2], [y1, y2], '-', color=arr_color, lw=lw)

def draw_dot(x, y):
    ax.plot(x, y, 'o', color=arr_color, markersize=4, zorder=5)

# ---------- 坐标 ----------
x_ref, y_m = 1.0, 5.0      # 参考输入
x_e,   _   = 3.0, 5.0      # 比较点
x_fork,_   = 4.0, 5.0      # 分叉点
x_pid, y_p = 5.5, 7.2      # P 分支
_,     y_i = 5.5, 5.0      # I 分支
_,     y_d = 5.5, 2.8      # D 分支
x_s2,  _   = 8.0, 5.0      # PID 求和点
x_pl,  _   = 10.5, 5.0     # 被控对象
x_fb0, _   = 12.0, 5.0     # 输出分叉
x_out, _   = 13.0, 5.0     # 输出
y_fb  = 7.5                # 反馈线高度
bw, bh = 0.55, 0.28        # 方框半宽半高

# ---------- 1. 方框 ----------
draw_box(x_pid, y_p, bw*2.2, bh*2)    # P
draw_box(x_pid, y_i, bw*2.2, bh*2)    # I
draw_box(x_pid, y_d, bw*2.2, bh*2)    # D
draw_box(x_pl, y_m, bw*2.2, bh*2.2, '#F4E8D0', '#8B6914')  # 被控对象
draw_box(x_pl, y_fb, bw*1.2, bh*1.4, sum_color, box_edge)   # 传感器

# ---------- 2. 圆圈 (比较点/求和点) ----------
draw_circle(x_e, y_m, 0.16)
draw_circle(x_s2, y_m, 0.18)

# ---------- 3. 文本 ----------
ax.text(x_pid, y_p, r'$K_p$', ha='center', va='center', fontsize=14)
ax.text(x_pid, y_i, r'$K_i\!\int$', ha='center', va='center', fontsize=13)
ax.text(x_pid, y_d, r'$K_d\frac{d}{dt}$', ha='center', va='center', fontsize=13)
ax.text(x_pl, y_m, '被控对象', ha='center', va='center', fontsize=11)
ax.text(x_pl, y_fb, '传感器', ha='center', va='center', fontsize=10)

ax.text(1.75, y_m+0.45, r'$r(t)$', ha='center', fontsize=13)       # 参考输入标签
ax.text(0.85, y_m-0.85, '参考输入', ha='center', fontsize=10)
ax.text(3.35, y_m+0.50, r'$e(t)$', ha='center', fontsize=13)       # 偏差
ax.text(8.85, y_m+0.50, r'$u(t)$', ha='center', fontsize=13)       # 控制量
ax.text(12.2, y_m+0.50, r'$y(t)$', ha='center', fontsize=13)       # 输出
ax.text(x_out+0.25, y_m-0.85, '输出', ha='center', fontsize=10)
ax.text(0.85, y_fb-0.60, '反馈', ha='center', fontsize=10)

# 比较点正负号
ax.text(x_e-0.50, y_m-0.38, r'$+$', fontsize=14, ha='center')
ax.text(x_e-0.35, y_m+0.45, r'$-$', fontsize=16, ha='center')
ax.text(x_s2-0.05, y_m+0.95, r'$+$', fontsize=12, ha='center')

# ---------- 4. 主通路连线 ----------
# r → 比较点
draw_arrow(x_ref, y_m, x_e-0.16, y_m)
# 比较点 → 分叉
draw_arrow(x_e+0.16, y_m, x_fork, y_m)
# 分叉 → P
draw_line(x_fork, y_m, x_fork, y_p)
draw_arrow(x_fork, y_p, x_pid-bw*1.1, y_p)
# 分叉 → I
draw_line(x_fork, y_m, x_pid-bw*1.1, y_m)
# 分叉 → D
draw_line(x_fork, y_m, x_fork, y_d)
draw_arrow(x_fork, y_d, x_pid-bw*1.1, y_d)
draw_dot(x_fork, y_m)

# P → 求和点
draw_arrow(x_pid+bw*1.1, y_p, x_s2-0.18, y_p)
draw_line(x_s2-0.18, y_p, x_s2-0.18, y_m+0.18)
draw_dot(x_s2-0.18, y_p)
# I → 求和点
draw_arrow(x_pid+bw*1.1, y_i, x_s2-0.18, y_i)
# D → 求和点
draw_arrow(x_pid+bw*1.1, y_d, x_s2-0.18, y_d)
draw_line(x_s2-0.18, y_d, x_s2-0.18, y_m-0.18)
draw_dot(x_s2-0.18, y_d)

# 求和点 → 被控对象
draw_arrow(x_s2+0.18, y_m, x_pl-bw*1.1, y_m)
# 被控对象 → 输出
draw_arrow(x_pl+bw*1.1, y_m, x_out, y_m)

# ---------- 5. 反馈通路 ----------
draw_dot(x_fb0, y_m)
draw_line(x_fb0, y_m, x_fb0, y_fb)
draw_arrow(x_fb0, y_fb, x_pl+bw*0.6, y_fb)       # → 传感器
draw_arrow(x_pl-bw*0.6, y_fb, 2.2, y_fb)          # ← 传感器
draw_line(2.2, y_fb, 2.2, y_m)
draw_arrow(2.2, y_m, x_e-0.16, y_m)

# ---------- 6. PID 虚线框 ----------
x_l, x_r = x_fork-0.22, x_s2+0.35
y_t, y_b = y_p+bh+0.25, y_d-bh-0.30
ax.plot([x_l, x_r, x_r, x_l, x_l], [y_t, y_t, y_b, y_b, y_t],
        '--', color=dash_color, lw=1.0)
ax.text(x_pid, y_t+0.30, 'PID 控制器', ha='center', fontsize=11, color='#333333')

# ---------- 7. 保存 ----------
plt.tight_layout(pad=0.5)
out_path = r'D:\bylw\code\lunwen\QLULatex\QLUThesisLatexTemplate-master\Thesis\static\figures\pid.png'
fig.savefig(out_path, dpi=200, bbox_inches='tight', facecolor='white', edgecolor='none')
print(f'PID 原理图已保存: {out_path}')
plt.close()
